File: 2018/Python/day08.py
# Advent of Code 2018 - Day 8

# Part 1

with open('day08.txt') as f:
    data = list(int(i) for i in f.read().split())


def find_metadata(nums):
    children = nums[0]
    n_metadata = nums[1]
    total = 0
    while children:
        meta, sub_nums = find_metadata(nums[2:])
        total += meta
        nums = nums[:2] + sub_nums
        children -= 1
    return total + sum(nums[2:2 + n_metadata]), nums[2 + n_metadata:]


print(find_metadata(data))

# Part 2


def find_value(nums):
    children = nums[0]
    n_metadata = nums[1]
    value = 0
    cs = [0]
    if not children:
        return sum(nums[2:2 + n_metadata]), nums[2 + n_metadata:]
    while children:
        val, sub_sums = find_value(nums[2:])
        cs.append(val)
        nums = nums[:2] + sub_sums
        children -= 1
    for v in nums[2:2 + n_metadata]:
        if v >= len(cs):
            continue
        value += cs[v]
    return value, nums[2 + n_metadata:]


print(find_value(data))

File: 2018/Haskell/src/Day08.hs
module Day08 where

import qualified Text.Parsec    as P
import Data.List.Extra
import Control.Monad.State
import Data.Either (fromRight)
import Control.Lens
import Data.Maybe (mapMaybe)

type Parser = P.Parsec [Int] ()

parseInput :: String -> [Int]
parseInput = map read . splitOn " " . init

sumMeta :: Parser Int
sumMeta = do
  nChildren <- P.anyToken
  nMeta <- P.anyToken
  cs <- sum <$> replicateM nChildren sumMeta
  ms <- sum <$> replicateM nMeta P.anyToken
  pure $ cs + ms

sumRoot :: Parser Int
sumRoot = do
  nChildren <- P.anyToken
  nMeta <- P.anyToken
  cs <- replicateM nChildren sumRoot
  ms <- replicateM nMeta P.anyToken
  pure $ case cs of
    [] -> sum ms
    _ -> sum $ mapMaybe (\x -> cs ^? ix (x - 1)) ms

part1 :: [Int] -> Int
part1 = fromRight 0 . P.parse sumMeta ""

part2 :: [Int] -> Int
part2 = fromRight 0 . P.parse sumRoot ""

-- Do it with state instead of Parser, much worse ...
sumState :: State [Int] Int
sumState = do
  xs <- get
  let nChild = head xs
  let nMeta = xs !! 1
  put $ drop 2 xs
  children <- sum <$> replicateM nChild sumState
  new <- get
  let meta = sum (take nMeta new)
  put $ drop nMeta new
  pure $ children + meta

sumStateMeta :: State [Int] Int
sumStateMeta = do
  xs <- get
  let nChild = head xs
  let nMeta = xs !! 1
  put $ drop 2 xs
  children <- replicateM nChild sumStateMeta
  new <- get
  let meta = take nMeta new
  put $ drop nMeta new
  pure $ if null children
            then sum meta
            else sum $ mapMaybe (\x -> children ^? ix (x - 1)) meta
----

main :: IO ()
main = do
  input <- parseInput <$> readFile "../data/day08.in"
  print $ part1 input
  print $ part2 input
  print $ runState sumState input
  print $ runState sumStateMeta input

-- 41028
-- 20849
